File: app/src/main/java/com/innovation4you/napking/ui/adapter/ListAdapter.java
package com.innovation4you.napking.ui.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

public abstract class ListAdapter<T, VH extends BaseViewHolder> extends RecyclerView.Adapter<VH> {

	final protected List<T> items;
	final protected View.OnClickListener onClickListener;

	public ListAdapter(final List<T> items, final View.OnClickListener onClickListener) {
		this.items = items;
		this.onClickListener = onClickListener;
	}

	@Override
	public VH onCreateViewHolder(ViewGroup parent, int viewType) {
		final View view = LayoutInflater.from(parent.getContext()).inflate(
				getLayoutResId(), parent, false);
		return onCreateViewHolder(view);
	}

	@Override
	public void onBindViewHolder(VH holder, int position) {
		onBindViewHolder(holder, items.get(position), position);
	}

	protected abstract int getLayoutResId();

	public abstract VH onCreateViewHolder(final View view);

	public abstract void onBindViewHolder(VH viewHolder, T item, int position);

	@Override
	public int getItemCount() {
		return items.size();
	}
}

File: app/src/main/java/com/innovation4you/napking/ui/adapter/ListAdapter.kt
package com.innovation4you.napking.ui.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class ListAdapter<T, VH : BaseViewHolder>(protected val items: List<T>, protected val onClickListener: View.OnClickListener) : RecyclerView.Adapter<VH>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val view = LayoutInflater.from(parent.context).inflate(
                layoutResId, parent, false)
        return onCreateViewHolder(view)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        onBindViewHolder(holder, items[position], position)
    }

    protected abstract val layoutResId: Int

    abstract fun onCreateViewHolder(view: View): VH

    abstract fun onBindViewHolder(viewHolder: VH, item: T, position: Int)

    override fun getItemCount(): Int {
        return items.size
    }
}
